feat(pfCard): Add a loading state to the pfCards

amarie401 opened the pull request:
## Description
Adding a loading state feature to the remaining pfCards as a continuation of PR #661. This would close #662 

[pfAggregateStatusCard](http://rawgit.com/amarie401/angular-patternfly/pf-cards-loading-dist/dist/docs/index.html#/api/patternfly.card.component:pfAggregateStatusCard)
[pfInfoStatusCard](http://rawgit.com/amarie401/angular-patternfly/pf-cards-loading-dist/dist/docs/index.html#/api/patternfly.card.component:pfInfoStatusCard)
[pfCards - Trends](http://rawgit.com/amarie401/angular-patternfly/pf-cards-loading-dist/dist/docs/index.html#/api/patternfly.card.component:pfCard%20-%20Trends) (example)

![loading-info](https://user-images.githubusercontent.com/20052391/32381013-006805ba-c088-11e7-828f-e9a17768148e.gif)

![loading-agg](https://user-images.githubusercontent.com/20052391/32381091-3c94d284-c088-11e7-938e-c1f822ff4d72.gif)

![loading-trends](https://user-images.githubusercontent.com/20052391/32381186-7cb27cc2-c088-11e7-95cd-e225da93e179.gif)


## PR Checklist

- [X] Unit tests are included
- [X] Screenshots are attached (if there are visual changes in the UI)
- [X] A Designer (@beanh66) is assigned as a reviewer (if there are visual changes in the UI)
- [x] A CSS rep (@cshinn) is assigned as a reviewer (if there are visual changes in the UI)

cshinn commented:
@amarie401 what's the -22px position on the spinner container for? to compensate for default padding or something? If so, would it be possible to use a pf variable that corresponds to the thing that's being compensated for instead?

dtaylor113 commented:
> @amarie401 what's the -22px position on the spinner container for?

Hi @cshinn, when card is in the process of loading, it's body/contents are not shown, so there is no 'height' to the card.  These changes add a height and adjust the position of the spinner.

jeff-phillips-18 commented:
Wondering if, at minimum, we should all the consumer to set the card height when loading. When these cards are in a dashboard, the size changes may cause issues.

beanh66 commented:
+ 1 @jeff-phillips-18 I was wondering the same thing, if it would be a problem when card heights all of a sudden change on a dashboard.

dtaylor113 commented:
> Wondering if, at minimum, we should all the consumer to set the card height when loading

Ok, we could add a `defaultCardHeight` or a `loadingCardHeight` to the main `status` configuration obj.

amarie401 commented:
I have added a loadingCardHeight option

dtaylor113 commented on src/card/aggregate-status/aggregate-status-card.component.js:
```diff
@@ -63,13 +66,19 @@
        <pf-aggregate-status-card status="aggStatusAlt" show-top-border="true" alt-layout="true"></pf-aggregate-status-card>
+       <br/>
+       <label>Loading State</label>
+       <pf-aggregate-status-card status="aggStatusAlt2" loading-card-height="150px" show-top-border="true" show-spinner="dataLoading" spinner-text="Loading" layout="tall"></pf-aggregate-status-card>
```
Looking around at other examples of this in angular-patternfly we do not include the 'px' as part of the string, just the number.

amarie401 replied:
Can remove :)

dtaylor113 commented on src/card/aggregate-status/aggregate-status-card.html:
```diff
@@ -28,8 +35,8 @@ <h2 class="card-pf-title">
 </div>
-<div ng-if="$ctrl.isMiniLayout" class="card-pf card-pf-aggregate-status card-pf-aggregate-status-mini" ng-class="{'card-pf-accented': $ctrl.shouldShowTopBorder}">
-  <h2 class="card-pf-title">
+<div ng-if="$ctrl.isMiniLayout" class="card-pf card-pf-aggregate-status card-pf-aggregate-status-mini" ng-class="{'card-pf-accented': $ctrl.shouldShowTopBorder}" ng-style="$ctrl.spinnerHeight">
```
Not sure why the ng-style was added here?  It seemed to work in the `show-spinner` div below.

dtaylor113 commented on src/card/aggregate-status/aggregate-status-card.html:
```diff
@@ -41,7 +48,14 @@ <h2 class="card-pf-title">
     </span>
   </h2>
-  <div class="card-pf-body">
+  <div class="card-pf-body" ng-class="{'show-spinner': $ctrl.showSpinner}">
```
Try moving the `ng-style` to here, it would be consistent with the `!mini` HTML above where we added the `ng-style`

dtaylor113 commented:
Looking good, although the loading icon doesn't appear to be centered:

infoStatusCard:
![image](https://user-images.githubusercontent.com/12733153/32464286-894ae168-c30d-11e7-8ead-45008ace131b.png)

aggregateStatusCard:
![image](https://user-images.githubusercontent.com/12733153/32464318-a09d92d4-c30d-11e7-888e-9cec2d4ecfed.png)

Seems like there is more space above the spinners than there is below the spinners.

amarie401 commented:
Okay will look into it!

jeff-phillips-18 commented on src/card/aggregate-status/aggregate-status-card.component.js:
```diff
@@ -30,6 +30,9 @@
  * @param {boolean=} show-top-border Show/hide the top border, true shows top border, false (default) hides top border
+ * @param {boolean=} showSpinner Show/Hide the spinner for loading state. True shows the spinner, false (default) hides the spinner
+ * @param {string=} spinnerText Text for the card spinner
+ * @param {string=} loadingCardHeight Height to set for the card when data is loading
```
We use loading here but showSpinner. I think it needs to be named such that it is obvious that this is the card height when showing the spinner.

dtaylor113 replied:
good point.   spinnerCardHeight?

amarie401 replied:
I think renaming to spinnerCardHeight is good to show the correlation

jeff-phillips-18 commented on src/card/aggregate-status/aggregate-status-card.html:
```diff
@@ -13,8 +13,15 @@ <h2 class="card-pf-title">
   </h2>
-  <div class="card-pf-body">
-    <p class="card-pf-aggregate-status-notifications">
+  <div class="card-pf-body" ng-class="{'show-spinner': $ctrl.showSpinner}" ng-style="$ctrl.spinnerHeight">
```
the spinnerHeight ought only be used when the spinner is shown.

amarie401 replied:
Agree

jeff-phillips-18 commented on src/card/aggregate-status/aggregate-status-card.html:
```diff
@@ -29,7 +36,7 @@ <h2 class="card-pf-title">
 </div>
 <div ng-if="$ctrl.isMiniLayout" class="card-pf card-pf-aggregate-status card-pf-aggregate-status-mini" ng-class="{'card-pf-accented': $ctrl.shouldShowTopBorder}">
-  <h2 class="card-pf-title">
+  <h2 class="card-pf-title" ng-class="{'hide-for-spinner': $ctrl.showSpinner}">
```
Should we be hiding the title while data is loading?

amarie401 replied:
I believe the title is showing and not hidden with the base pfCard from my previous PR, I can change this to be consistent.

amarie401 commented:
I have changed the name from loadingCardHeight to spinnerCardHeight for better correlation, unhid the title when loading and updated the ``ng-style`` to make the spinnerHeight more dynamic. @dtaylor113 are you still seeing the icon not being centered?

beanh66 commented:
@amarie401 did you update the gif? I am still seeing that it isn't quite centered but I might be looking at an old version!

amarie401 commented:
@beanh66 @dtaylor113 

They seem to be centered to be in their respective cards. Although the frame that I recorded the gif isn't exactly the same..Thoughts?

aggStatus
![loading-update1](https://user-images.githubusercontent.com/20052391/32508300-e0a18d72-c3b7-11e7-94e1-ca139479044f.gif)

infoStatus
![loading-update2](https://user-images.githubusercontent.com/20052391/32508317-ebf9191a-c3b7-11e7-9a4f-3472f013a2c3.gif)

jeff-phillips-18 commented:
The info status card should likely show the title and the icon just not the status information. The structure of the card, however, seems incorrect. The card-pf-title is inside the card-pf-body. Can we work around that and still show the title and image but replace the status info content with the spinner?

dtaylor113 commented:
> The info status card should likely show the title and the icon just not the status information.

Might the icon and/or title change based upon the loaded data? Ie. show an error icon/msg?  If so, we should just hide everything until the data is loaded.

jeff-phillips-18 commented:
The application can leave those values unset if that be the case. Can we try that scenario as well?

jeff-phillips-18 commented:
My point being, that a blank card with just a spinner gives no indication of what is being loaded. If any of the info can be shown, then it gives the user an idea of what they are waiting for.

dtaylor113 commented:
Went looking to find PF doc/examples/css/HTML for this card, I can't find InfoStatusCard in any of these locations:
https://github.com/patternfly/patternfly-design/tree/master/pattern-library/cards
http://www.patternfly.org/pattern-library/cards/base-card/
https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/card-view-single-select.html

jeff-phillips-18 commented:
@dtaylor113 The PR for this is still up in patternfly-design: https://github.com/patternfly/patternfly-design/pull/483

dtaylor113 commented:
> @dtaylor113 The PR for this is still up in patternfly-design: patternfly/patternfly-design#483

ah...that's why I couldn't find it :-)  So this seems to be a case of implementation before official PF html/css/less.

jeff-phillips-18 commented:
> So this seems to be a case of implementation before official PF html/css/less.
Yeah, feel free to implement in PF first 😉

dtaylor113 commented:
Still not exactly centered.  'with HTML' card doesn't initially set icon or title.  'Loading State' card  does initially set icon and title:

![loading](https://user-images.githubusercontent.com/12733153/32568538-8b28f18e-c48c-11e7-84a1-791844765a4f.gif)

amarie401 commented:
Does the icon have any margins? Seems like the icon is being pushed over

dtaylor113 commented:
Hi, pushed latest changes.  The infoStatusCard spinner icon is now centered within the card, with and without titles/icons:

![image](https://user-images.githubusercontent.com/12733153/32676280-09b2ceb6-c628-11e7-8623-81b1178fe1c7.png)

jeff-phillips-18 commented on src/card/aggregate-status/aggregate-status-card.component.js:
```diff
@@ -103,28 +112,49 @@
       ]
-     };
+    };
+
```
Should setup the $scope.aggStatusAlt2 here with the title but the count and notifications being unknown.

jeff-phillips-18 commented on src/card/aggregate-status/aggregate-status-card.html:
```diff
@@ -13,8 +13,15 @@ <h2 class="card-pf-title">
   </h2>
-  <div class="card-pf-body">
-    <p class="card-pf-aggregate-status-notifications">
+  <div class="card-pf-body" ng-class="{'show-spinner': $ctrl.showSpinner}" ng-style="$ctrl.showSpinner ? $ctrl.spinnerHeight : {}">
```
I would expect the height to be set on the card, not on the body. It is called spinnerCardHeight and given the title height could potentially change based on knowing the title prior to data load it would be better to set the full card height.

jeff-phillips-18 commented on src/card/info-status/info-status-card.html:
```diff
@@ -1,11 +1,10 @@
     <span class="info-icon {{$ctrl.status.iconClass}}"></span>
   </div>
-  <div class="card-pf-info-content">
+  <div class="card-pf-info-content card-pf-body" ng-class="{'show-spinner': $ctrl.showSpinner}" ng-style="$ctrl.showSpinner ? $ctrl.spinnerHeight : {}">
```
Again, the height should be on the entire card

dtaylor113 commented on src/card/examples/card-trend.js:
```diff
@@ -5,7 +5,9 @@
+ * @param {string=} spinnerText Text for the card spinner
  * @param {boolean=} showTopBorder Show/Hide the blue top border. True shows top border, false (default) hides top border
+ * @param {boolean=} showSpinner Show/Hide the spinner for loading state. True shows the spinner, false (default) hides the spinner
  * @param {boolean=} showTitlesSeparator Show/Hide the grey line between the title and sub-title.
```
Need to document `spinnerCardHeight`

amarie401 commented:
Now with the spinner height on the entire card

![13-agg](https://user-images.githubusercontent.com/20052391/32739420-cede07c2-c86d-11e7-80ec-432feb7c008c.gif)

amarie401 commented:
With @jeff-phillips-18 changes

aggStatus:

![jeff-changes-agg](https://user-images.githubusercontent.com/20052391/32745842-f8019902-c881-11e7-8f44-54166b3a05fb.gif)

infoStat:

![jeff-changes-infostat](https://user-images.githubusercontent.com/20052391/32745907-1fed4826-c882-11e7-9e2a-7e13b8491b14.gif)

jeff-phillips-18 commented:
Not sure why your image above doesn't have the white background for the aggregate status cards. I pulled and tested and they correctly had the white background.

amarie401 commented:
@jeff-phillips-18 I noticed that as well. The giphy recording tool I'm using is being a bit weird but the actual background is correct.
